lib/charms/tempo_coordinator_k8s/v0/tempo_api.py

```diff
@@ -0,0 +1,219 @@
+        self.framework.observe(self.on.some_event_that_changes_tempos_url, self.do_something_to_publish)
+```
+"""
+
+import json
+import logging
+from typing import Optional
+
+from ops import Application, RelationMapping
+from pydantic import AnyHttpUrl, BaseModel, Field
+
+# The unique Charmhub library identifier, never change it
+LIBID = "6d55454c9a104113b2bd01e738dd5f99"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+PYDEPS = ["pydantic>=2"]
+
+log = logging.getLogger(__name__)
+
+DEFAULT_RELATION_NAME = "tempo-api"
+
+
+class TempoApiAppData(BaseModel):
+    """Data model for the tempo-api interface."""
+
+    ingress_url: Optional[AnyHttpUrl] = Field(
+        default=None,
+        description="The non-internal URL at which this application can be reached.  Typically, this is an ingress URL.",
+    )
+    direct_url: AnyHttpUrl = Field(
+        description="The cluster-internal URL at which this application can be reached.  Typically, this is a"
+        " Kubernetes FQDN like name.namespace.svc.cluster.local for connecting to the prometheus api"
+        " from inside the cluster, with scheme."
+    )
+
+
+class TempoApiRequirer:
+    """The requirer side of the tempo-api relation."""
+
+    def __init__(
+        self,
+        relations: RelationMapping,
+        relation_name: str = DEFAULT_RELATION_NAME,
+    ) -> None:
+        """Initialize the TempoApiRequirer object.
+
+        This object is for accessing data from relations that use the tempo-api interface.  It **does not**
+        autonomously handle the events associated with that relation.  It is up to the charm using this object to
+        observe those events as they see fit.  Typically, that charm should observe this relation's relation-changed
+        event.
+
+        This object is for interacting with a relation that has limit=1 set in charmcraft.yaml.  In particular, the
+        get_data method will raise if more than one related application is available.
+
+        Args:
+            relations: The RelationMapping of a charm (typically `self.model.relations` from within a charm object).
+            relation_name: The name of the relation.
+        """
+        self._charm_relation_mapping = relations
+        self._relation_name = relation_name
+
+    @property
+    def relations(self):
+        """Return the relation instances for applications related to us on the monitored relation."""
+        return self._charm_relation_mapping.get(self._relation_name, ())
+
+    def get_data(self) -> Optional[BaseModel]:
+        """Return data for at most one related application, raising if more than one is available.
+
+        Useful for charms that always expect exactly one related application.  It is recommended that those charms also
```

Comment:
I don't see the implementation for the limit=N case; what are you supposed to do then?

Reply:
atm, there is none.  Only the limit=1 case is provided

Reply:
This should be clearer in the docstrings so I'm resolving it.  There's another open thread about this same thing so if that has an issue I'll reopen this one too